Document existing indented syntax (#3969)
sass/language: spec/statement.md

# Statement

## Table of Contents

* [Definitions](#definitions)
  * [Current indentation level](#current-indentation-level)
  * [Document indentation character](#document-indentation-character)
* [Syntax](#syntax)
  * [ScssStatements](#scssstatements)
  * [IndentedStatements](#indentedstatements)
  * [Statements](#statements)
  * [Block](#block)
  * [Whitespace](#whitespace)
  * [Indentation](#indentation)

## Definitions

### Current indentation level

The `current indentation level` is the count of the [`documentation indentation
character`] between a new line and any non-whitespace characters of the last

> consumed line.

Reviewer: This should explicitly exclude lines that are only whitespace from consideration.

Author: I updated this by adding a second sentence. Alternatively, I considered-

```md
The `current indentation level` is the count of the [`documentation indentation
character`] immediately preceding the start of the last consumed statement and 
after the closest linebreak.
```

I also excluded lines that do not start a statement. The only case I can think where this currently applies is with multiline comments, but it will be useful to have already defined.

A document parsed in the indented syntax may not begin with whitespace, and the
initial `current indentation level` for any document is `0`.

Changes in the indentation level are used by the indented syntax to start and

> end blocks of statements.

Reviewer: This should be in a block quote to indicate that it's descriptive rather than normative.

[`documentation indentation character`]: #document-indentation-character

### Document indentation character

The `document indentation character` is the character used for calculating the

> [`current indentation level`], and is either a space or tab character.

Reviewer: The definition should come before the non-normative description that explains why the definition is useful, and the description should be in a block quote.

[`current indentation level`]: #current-indentation-level

> It is defined as the first tab or space character used as an [IndentMore]

Reviewer: ```suggestion
It is defined as the first tab or space character used as an [`IndentMore`]
```

production in a document.

In the indented syntax, no character other than the `document indentation
character` may be used for indentation.

## Syntax

### ScssStatements

<x><pre>

> **ScssStatements**      ::= (Statement ';'?¹)*Statement?

Reviewer: ```suggestion
**ScssStatements**      ::= (Statement ';'?¹)* Statement?
```

</pre></x>

1: This production is mandatory unless the previous `Statement` is a
`LoudComment` or `SilentComment`, or after the final production in a
`ScssStatements` production, which is either at the end of a `Block` or the end

> of the document.

Reviewer: You don't need to explicitly exempt the final statement—that's why the BNF ends in `Statement?` without a semicolon. You *do* need to explicitly exempt `';` after a `Statement` that ends in a `Block`, or else `a {b {} c: d}` would be invalid.

If a `WhitespaceComment` would be ambiguous with a `Statement` in the
`ScssStatements` rule, parse it preferentially as a `Statement`.

### IndentedStatements

<x><pre>
**IndentedStatements**  ::= (Statement IndentedSame?¹)* Statement?
</pre></x>

1: This production is mandatory unless the previous `Statement` is a
`LoudComment` or `SilentComment`, or after the final production in a
`IndentedStatements` production, which is either at the end of a `Block` or the

> end of the document.

Reviewer: This is _not_ true for the indented syntax. If you could leave out `IndentedSame`, then `/**/ a: b` would parse as two statements in sequence rather than an error as it currently does. It's also not true that `Statement IndentedSame` is a valid `IndentedStatements`—that would consume the indentation without consuming a statement afterwards, leaving the parser in a confusing and likely broken state.

If a `WhitespaceComment` would be ambiguous with a `Statement` in the

> `IndentedStatements` rule, parse it preferentially as a `Statement`.

Reviewer: How would this be ambiguous in the indented syntax?

### Statements

<x><pre>

> **Statements**          ::= [ScssStatements] | [IndentedStatements]¹

Reviewer: Where you had the footnote, it looks like it's just associated with `IndentedStatements` and not `ScssStatements`. Another alternative would be to add parentheses to make the grouping explicit.

```suggestion
**Statements**¹          ::= [ScssStatements] | [IndentedStatements]
```

</pre></x>

[ScssStatements]: #scssstatements
[IndentedStatements]: #indentedstatements

Only the production for the current syntax is valid.

### Block

<x><pre>
**ScssBlock**      ::= '{' [ScssStatements] '}'
**IndentedBlock**  ::= [IndentMore] [IndentedStatements] [IndentLess]
**Block**          ::= ScssBlock | IndentedBlock¹
</pre></x>

[IndentMore]: #indentation
[IndentLess]: #indentation

1: Only the production for the current syntax is valid.

> ### Whitespace

Reviewer: What about comments? What about a `Whitespace` production?

Author: Added in 70b12049d7bc462463d05af4dca8419ad98a4100

> Whitespace separates productions inside a statement when disambiguation is

> > necessary. The Whitespace productions do not separate Statements.

Reviewer: Several issues here:

* This seems like it's supposed to be normative, but it's in a block quote which indicates descriptive text.
* What does "inside a statement" mean? Block productions and thus the `Statements` production are also inside statements.
* When is disambiguation necessary? Does this mean that whitespace isn't allowed any time it wouldn't strictly disambiguate something?
* Statements certainly _are_ separated by whitespace in SCSS. `@foo; @bar` is a valid stylesheet with whitespace between the statements.

Author: This was indeed meant to be a non-normative comment to remind myself to consider the difference between subtraction and unary negation, for example, and the difference between Whitespace and Indentation. It's not needed anymore.

<x><pre>
**LineBreak**               ::= CarriageReturn | LineFeed | FormFeed
**ScssWhitespace**          ::= LineBreak | Space | Tab
**IndentedWhitespace**      ::= Space | Tab
</pre></x>

### Indentation

<x><pre>
**IndentCharacter**         ::= Space | Tab
**IndentSame**              ::= [LineBreak] [IndentCharacter]{ Current }

> **IndentLess**              ::= [LineBreak] [IndentCharacter]{ 0, Current - 1 }

Reviewer: If this consumes the indentation, then there won't be anything for the next statement to consume as `IndentSame`. The end of the document should also be considered an `IndentLess`.

Author: Definitely check my logic on this, but I think I can remove the `IndentLess` production. It was only used in the Comments and `IndentedBlock` productions, and I can remove it because it was essentially operating as a look ahead, and wasn't necessary. 

In `IndentedBlock`, the `IndentedStatements` production will consume until there is no longer an  `IndentSame`, so we don't need to explicitly end it with `IndentLess`. I'm less confident on the Indented comment productions ending with a wildcard.

> **IndentMore**              ::= [LineBreak] [IndentCharacter]{ Current + 1, }

Reviewer: ```suggestion
**IndentMore**              ::= [LineBreak] [IndentCharacter]{ Current + 1 }
```

Author: I have the trailing comma to denote that there can be 1 or more `IndentCharacter`.

> **IndentMore**              ::= [LineBreak] [IndentCharacter]{ Current + 1, }

Reviewer: These should all allow arbitrarily many whitespace-only lines.

</pre></x>

[LineBreak]: #whitespace
[IndentCharacter]: #whitespace

> The [IndentCharacter] must be the [Document indentation character].

Reviewer: ```suggestion
The [IndentCharacter] must be the [document indentation character].
```

[Document indentation character]: #document-indentation-character

`Current` is the [current indentation level] for a document. After consuming an
`IndentSame`, `IndentLess`, or `IndentMore` production, the [current indentation
level] is set to the number of [IndentCharacter]s found. At the end of a

> document, the [current indentation level] is set to 0.

Reviewer: Why?

Author: It was an incorrect attempt to trigger an `IndentLess` production.

# Statement

## Table of Contents

* [Definitions](#definitions)
  * [Current indentation level](#current-indentation-level)
  * [Document indentation character](#document-indentation-character)
* [Syntax](#syntax)
  * [ScssStatements](#scssstatements)
  * [IndentedStatements](#indentedstatements)
  * [Statements](#statements)
  * [Block](#block)
  * [Comments](#comments)
    * [LoudComment](#loudcomment)
    * [SilentComment](#silentcomment)
    * [WhitespaceComment](#whitespacecomment)
  * [Whitespace](#whitespace)
  * [Indentation](#indentation)

## Definitions

### Current indentation level

The `current indentation level` is the count of the [`documentation indentation

> character`] between a new line and any non-whitespace characters of the last

Reviewer: "the last consumed statement" is a good way to indicate this :+1:.

consumed statement. Lines that only contain whitespace or do not start a
statement do not impact the `current indentation level`.

A document parsed in the indented syntax may not begin with whitespace, and the
initial `current indentation level` for any document is `0`.

> Changes in the indentation level are used by the indented syntax to start and
> end blocks of statements.

[`documentation indentation character`]: #document-indentation-character

### Document indentation character

The `document indentation character` is defined as the first tab or space

> character used as an [`IndentMore`] production in a document.

Reviewer: ```suggestion
character used in an [`IndentMore`] production in a document.
```

since indentation is likely to use multiple spaces.

[`IndentMore`]: #indentation

> The `document indentation character` is the character used for calculating the
> [`current indentation level`]. In the indented syntax, no character other than
> the `document indentation character` may be used for indentation.

[`current indentation level`]: #current-indentation-level

## Syntax

### ScssStatements

<x><pre>
**ScssStatements**      ::= (Statement ';'?¹)* Statement?
</pre></x>

1: This production is mandatory unless the previous `Statement` is a
`LoudComment`, `SilentComment`, or ends in a `Block`.

If a `WhitespaceComment` would be ambiguous with a `Statement` in the
`ScssStatements` rule, parse it preferentially as a `Statement`.

> ### IndentedStatements

Reviewer: This should include a requirement that the statements not include newlines outside of `IndentSame` productions, or else it's halfway to the more flexible syntax already.

<x><pre>
**IndentedStatements**  ::= (Statement [IndentSame])* Statement
</pre></x>

[IndentSame]: #indentation

### Statements

<x><pre>

> **Statements**          ::= ([ScssStatements] | [IndentedStatements])¹

Reviewer: Let's call this `Stylesheet` to emphasize that it's the root production of the document. It should also allow an optional U+FEFF at the beginning as a [byte-order mark](https://en.wikipedia.org/wiki/Byte_order_mark).

</pre></x>

[ScssStatements]: #scssstatements
[IndentedStatements]: #indentedstatements

1. Only the production for the current syntax is valid.

### Block

<x><pre>
**ScssBlock**      ::= '{' [ScssStatements] '}'
**IndentedBlock**  ::= [IndentMore] [IndentedStatements]
**Block**          ::= (ScssBlock | IndentedBlock)¹
</pre></x>

[IndentMore]: #indentation

1: Only the production for the current syntax is valid.

### Comments

#### LoudComment

<x><pre>

> **ScssLoudComment**          ::= '/*' '!'? (.* | Interpolation)*'*/'

Reviewer: You don't need explicit `'!'?` here, since it's covered by the general production. You _do_ need to make sure the `.*` isn't allowed to consume `#{` or `*/`, though. (Similarly below.)

Also, this is a style nit but it affects readability pretty strongly, make sure to have spaces between your productions:

```suggestion
**ScssLoudComment**          ::= '/*' '!'? (.* | Interpolation)* '*/'
```

> **IndentedLoudComment**      ::= '/*' '!'? (.* | Interpolation)*

Reviewer: Similarly, this `.*` needs to indicate that it can't contain newlines. (Similarly below.)

Also: statement-level comments in the indented syntax allow nested content. (It's weird, I know.)

**LoudComment**              ::= (ScssLoudComment | IndentedLoudComment)¹
</pre></x>

> > TODO: Wildcard and character set regex syntax conflict with markdown.

Author: If the `.*` syntax is what's needed here, it is messing with markdown, and I wasn't sure if it should be correct BNF or correctly rendered markdown in that case.

I also considered using `'/*' '!'? ([^IndentationLess] | Interpolation)*` to help narrow when the comment production ends without consuming the indentation. The `[]` also conflicted with markdown syntax.

Reviewer: You can avoid markdown parsing by putting a backslash before the character. For example, `.\* .\*` renders as ".\* .\*" and `\[foo\]` renders as \[foo\].

1: Only the production for the current syntax is valid.

#### SilentComment

<x><pre>
**ScssSilentComment**          ::= '//' .*
**IndentedSilentComment**      ::= '//' .*
**SilentComment**              ::= (ScssSilentComment | IndentedSilentComment)¹
</pre></x>

1: Only the production for the current syntax is valid.

#### WhitespaceComment

<x><pre>

> **ScssWhitespaceComment**          ::= '//' .*'//' | '/*' .*'*/'

Reviewer: Why the trailing `'//'`?

**IndentedWhitespaceComment**      ::= '/*' .* '*/'
**WhitespaceComment*¹              ::= ScssWhitespaceComment
&#32;                                | IndentedWhitespaceComment
</pre></x>

1: Only the production for the current syntax is valid.

### Whitespace

<x><pre>
**LineBreak**               ::= CarriageReturn | LineFeed | FormFeed
**ScssWhitespace**          ::= LineBreak | Space | Tab
**IndentedWhitespace**      ::= Space | Tab
**Whitespace**              ::= (ScssWhitespace | IndentedWhitespace)¹
&#32;                         | [WhitespaceComment]
</pre></x>

1: Only the production for the current syntax is valid.

[WhitespaceComment]: #whitespacecomment

### Indentation

<x><pre>

> **WhitespaceOnlyLine**          ::= IndentSame (Whitespace)*

Reviewer: This shouldn't require `IndentSame`. It's valid to have empty zero-indent lines between indented lines.

It should also require a `LineBreak` at the end.

**IndentCharacter**             ::= Space | Tab

> **IndentSame**                  ::= WhitespaceOnlyLine* [LineBreak]

Reviewer: `WhitespaceOnlyLine*` should come after `LineBreak`; otherwise you're parsing trailing whitespace after the last statement as "a line".

&#32;                               [IndentCharacter]{ Current }
**IndentMore**                  ::= WhitespaceOnlyLine* [LineBreak]

> &#32;                               [IndentCharacter]{ Current + 1, }

Reviewer: It might be clearer to write this

```suggestion
&#32;                               [IndentCharacter]{ ≥ Current + 1 }
```

### Block

<x><pre>
**ScssBlock**      ::= '{' [ScssStatements] '}'
**IndentedBlock**  ::= [IndentMore] [IndentedStatements]
**Block**          ::= (ScssBlock | IndentedBlock)¹
</pre></x>

[IndentMore]: #indentation

1: Only the production for the current syntax is valid.

### Comments

#### LoudComment

<x><pre>
**ScssLoudComment**          ::= '/\*' (.\*¹ | Interpolation)\* '\*/'
**InterpolatedCommentText**  ::= (.\*² | Interpolation)\*
**IndentedLoudChildren**     ::= (InterpolatedCommentText [IndentSame])\*
&#32;                            InterpolatedCommentText
**IndentedLoudComment**      ::= '/\*' InterpolatedCommentText
&#32;                            ([IndentMore] IndentedLoudChildren)?
**LoudComment**              ::= (ScssLoudComment | IndentedLoudComment)³
</pre></x>

1. This may not contain `#{` or `*/`.
2. This may not contain `#{` or newlines.
3. Only the production for the current syntax is valid.

#### SilentComment

<x><pre>
**ScssSilentComment**          ::= '//' .\*¹
**CommentText**                ::= .\*²
**IndentedSilentChildren**     ::= (CommentText [IndentSame])\* CommentText
**IndentedSilentComment**      ::= '//' CommentText ([IndentMore]

> &#32;                              IndentedSilentChildren)?

Reviewer: Don't split parentheses across lines like this.

```suggestion
**IndentedSilentComment**      ::= '//' CommentText
&#32;                              ([IndentMore] IndentedSilentChildren)?
```

**SilentComment**              ::= (ScssSilentComment | IndentedSilentComment)³
</pre></x>

1. This may not contain newlines.

> 2. This may not contain newlines outside of [IndentSame] productions.

Reviewer: How would `CommentText` contain an `IndentSame` production?

Author: Good point. I changed it to `[WhitespaceOnlyLine]` because [silent comments consume](https://sass-lang.com/playground/#eJwVizEOwjAQBPt7xZZQBdr8xlwWfJJzjuyLIL/HLqYYaebxXBYRIA2A1wqdFpnQuu/0gCaHuZZzI77Zgv1ISlQvF4o5++hvc+iXR/oh2yeXQbDB+lxra9S4i8g2Wq54/wEWeCSu) WhitespaceOnlyLines as well.

Reviewer: The issue is that this doesn't contain *any* other productions. It's just matching pure characters. It should forbid newlines entirely.

Author: Ok, that makes sense. I forbade newlines in https://github.com/sass/sass/pull/3969/commits/1c19f21c9e0a9c5736dfc78d52839c10480af847.

character`] between a new line and any non-whitespace characters of the last
consumed statement. Lines that only contain whitespace or do not start a
statement do not impact the `current indentation level`.

A document parsed in the indented syntax may not begin with whitespace, and the
initial `current indentation level` for any document is `0`.

> Changes in the indentation level are used by the indented syntax to start and
> end blocks of statements.

[`documentation indentation character`]: #document-indentation-character

### Document indentation character

The `document indentation character` is defined as the first tab or space
character used in an [`IndentMore`] production in a document.

[`IndentMore`]: #indentation

> The `document indentation character` is the character used for calculating the
> [`current indentation level`]. In the indented syntax, no character other than
> the `document indentation character` may be used for indentation.

[`current indentation level`]: #current-indentation-level

## Syntax

### ScssStatements

<x><pre>
**ScssStatements**      ::= (Statement ';'?¹)* Statement?
</pre></x>

> 1. This production is mandatory unless the previous `Statement` is a

Reviewer: Use `1:` rather than `1.` for annotation references. Otherwise Markdown will render them as an ordered list.

Author: In sections with multiple annotation references, I kept the ordered list to avoid rendering it as a single paragraph. If there is a single reference, I changed it to `1:`.

Reviewer: Instead of this, please separate out the separate references into their own paragraphs, as [in the existing spec](https://github.com/sass/sass/blob/1c9ec00bbb061fa1be92c2e8c81144e6b8d8159b/spec/functions.md?plain=1#L66-L73)

Author: Updated in https://github.com/sass/sass/pull/3969/commits/63f009e0bd754eb37423a240e7b97dab6753d011

Author: Updated in https://github.com/sass/sass/pull/3969/commits/63f009e0bd754eb37423a240e7b97dab6753d011

`LoudComment`, `SilentComment`, or ends in a `Block`.

If a `WhitespaceComment` would be ambiguous with a `Statement` in the
`ScssStatements` rule, parse it preferentially as a `Statement`.

### IndentedStatements

<x><pre>
**IndentedStatements**  ::= (Statement [IndentSame])* Statement
</pre></x>

[IndentSame]: #indentation

The `Statement` productions may not include newlines outside of `IndentSame`
productions.

### Stylesheet

<x><pre>
**Stylesheet**          ::= U+FEFF? ([ScssStatements] | [IndentedStatements])¹
</pre></x>

[ScssStatements]: #scssstatements
[IndentedStatements]: #indentedstatements

1. Only the production for the current syntax is valid.

### Block

<x><pre>
**ScssBlock**      ::= '{' [ScssStatements] '}'
**IndentedBlock**  ::= [IndentMore] [IndentedStatements]
**Block**          ::= (ScssBlock | IndentedBlock)¹
</pre></x>

[IndentMore]: #indentation

1. Only the production for the current syntax is valid.

### Comments

#### LoudComment

<x><pre>
**ScssLoudComment**          ::= '/\*' (.\*¹ | Interpolation)\* '\*/'

> **IndentedLoudCommentNest**  ::= [IndentMore] ((.\*² | Interpolation)\*
> &#32;                            [IndentSame])\* (.\*² | Interpolation)\*
> **IndentedLoudComment**      ::= '/\*' (.\*² | Interpolation)\*
> &#32;                            IndentedLoudCommentNest?

Reviewer: I would write this as:

```suggestion
**InterpolatedCommentText**  ::= (.\*² | Interpolation)\*
**IndentedLoudChildren**     ::= (InterpolatedCommentText [IndentSame])\*
&#32;                            InterpolatedCommentText
**IndentedLoudComment**      ::= '/\*' InterpolatedCommentText
&#32;                            ([IndentMore] IndentedLoudChildren)?
```

This way you avoid repeating the interpolation construction over and over again, and `IndentedLoudChildren` production matches the structure of `IndentedStatements`.

**LoudComment**              ::= (ScssLoudComment | IndentedLoudComment)³
</pre></x>

1. This may not contain `#{` or `*/`.
2. This may not contain `#{` or newlines.
3. Only the production for the current syntax is valid.

#### SilentComment

<x><pre>
**ScssSilentComment**          ::= '//' .\*¹

> **IndentedSilentComment**      ::= '//' .\*²

Reviewer: This can also have text nested within it.

Author: Added in 22c1607cb6494b98dd6aeccf8da9b877234439f3

**SilentComment**              ::= (ScssSilentComment | IndentedSilentComment)³
</pre></x>

1. This may not contain newlines.
2. This may not contain newlines outside of [IndentSame] productions.
3. Only the production for the current syntax is valid.

#### WhitespaceComment

<x><pre>
**ScssWhitespaceComment**          ::= '//' .\*¹ | '/\*' .\*² '\*/'

> **IndentedWhitespaceComment**      ::= '/\*' .\*² '\*/'

Reviewer: The indented syntax also allows `//` as whitespace comments [at the end of a line](https://sass-lang.com/playground/#eJwzMHRISU0qTVdIy89X0NdXSM7PzU3NKwEAVhAHng==). Also, this should not allow newlines.

Author: Added in 22c1607cb6494b98dd6aeccf8da9b877234439f3

**WhitespaceComment**³             ::= ScssWhitespaceComment
&#32;                                | IndentedWhitespaceComment
</pre></x>

1. This may not contain newlines.
2. This may not contain `*/`.
3. Only the production for the current syntax is valid.

### Whitespace

<x><pre>

> **LineBreak**               ::= CarriageReturn | LineFeed | FormFeed
> **ScssWhitespace**          ::= LineBreak | Space | Tab
> **IndentedWhitespace**      ::= Space | Tab
> **Whitespace**              ::= (ScssWhitespace | IndentedWhitespace)¹
> &#32;                         | [WhitespaceComment]
> </pre></x>
> 
> 1. Only the production for the current syntax is valid.

Reviewer: Potentially a simpler way to define this would be:

```suggestion
**LineBreak**  ::= CarriageReturn | LineFeed | FormFeed
**Whitespace** ::= LineBreak¹ | Space | Tab | [WhitespaceComment]
</pre></x>

1: This is not allowed in the indented syntax.
```

# Statement

## Table of Contents

* [Definitions](#definitions)
  * [Current indentation level](#current-indentation-level)
  * [Document indentation character](#document-indentation-character)
* [Syntax](#syntax)
  * [ScssStatements](#scssstatements)
  * [IndentedStatements](#indentedstatements)
  * [Stylesheet](#stylesheet)
  * [Block](#block)
  * [Comments](#comments)
    * [LoudComment](#loudcomment)
    * [SilentComment](#silentcomment)
    * [WhitespaceComment](#whitespacecomment)
  * [Whitespace](#whitespace)
  * [Indentation](#indentation)

## Definitions

### Current indentation level

> The *current indentation level* is the number of [documentation indentation

Reviewer 2: shouldn't this be `document` instead of `documentation` ?

Author: Opened #3992 to fix. Thanks!

3: Only the production for the current syntax is valid.

#### SilentComment

<x><pre>
**ScssSilentComment**          ::= '//' .\*¹
**CommentText**                ::= .\*²
**IndentedSilentChildren**     ::= (CommentText [IndentSame])\* CommentText
**IndentedSilentComment**      ::= '//' CommentText
&#32;                              ([IndentMore] IndentedSilentChildren)?
**SilentComment**              ::= (ScssSilentComment | IndentedSilentComment)³
</pre></x>

1: This may not contain newlines.

2: This may not contain newlines outside of [WhitespaceOnlyLine] productions.

3: Only the production for the current syntax is valid.

[WhitespaceOnlyLine]: #indentation

#### WhitespaceComment

<x><pre>
**WhitespaceComment**              ::= ('//' .\*¹) | ('/\*' .\*² '\*/')
</pre></x>

1: This may not contain newlines.

2: This may not contain `*/`. In the indented syntax, this may not contain
   newlines.

### Whitespace

<x><pre>
**LineBreak**  ::= CarriageReturn | LineFeed | FormFeed
**Whitespace** ::= LineBreak¹ | Space | Tab | [WhitespaceComment]
</pre></x>

1: This is not allowed in the indented syntax.

[WhitespaceComment]: #whitespacecomment

### Indentation

<x><pre>

> **WhitespaceOnlyLine**          ::= IndentSame? [Whitespace]\* [LineBreak]

Reviewer: There's not really a reason to have `IndentSame?` here. The point is that empty lines can have any whitespace and have no bearing on the current indentation level.
